server/src/main/java/io/druid/client/CachingClusteredClient.java

```diff
@@ -623,7 +772,41 @@ private void addSequencesFromServer(
             }
             return res;
           })
-          .flatMerge(seq -> seq, query.getResultOrdering());
+          .flatMerge(Function.identity(), query.getResultOrdering());
+    }
+  }
+
+  // POJO
+  private static class ServerMaybeSegmentMaybeCache<T>
+  {
+    private final DruidServer server;
+    private final Optional<SegmentDescriptor> segmentDescriptor;
```

Comment:
Optional stored in field (as well as Map.Entry or Map value, above in this class) [is considered a code smell](https://stackoverflow.com/a/26328314/648955). IMO, this class overuses Optionals to some degree, as well as Streams. Replacing some of them with nullable variables will only make the code clearer

Reply:
the reason some of these use Optional is things like Streams don't like grouping by nullable values, so having an Optional still allows grouping handling for keys in maps that would otherwise not handle null keys.

Reply:
I don't see where Stream grouping is done by segmentDescriptor or cachedValue of this POJO

Reply:
I think I fixed this, the grouping thing must have been an older implementation